tests/testScrapPaper.js

```diff
@@ -0,0 +1,386 @@
+      });
+      it('database barcodes should checksum sha16 successfully against base36 barcode components', () => {
+        barcodeRows.forEach((row) => {
+          const { base36, sha16 } = getBarcodeSegments(row.barcode);
+          const recomputedSha16 = jsCrc.crc16(base36);
+          assert.equal(recomputedSha16, sha16);
+        });
+      });
+      // it('barcodes should ALL be scannable by barcode reader', async () => {
+      //   decodedJpegs = await decodeBarcodes(
+      //     pdfBuffer,
+      //     `${config.imageProcessingDir}/test`,
+      //     'Any original filename.pdf',
+      //     console
+      //   );
+      //   assert.isDefined(decodedJpegs);
+      //   decodedJpegs.forEach((jpeg) => {
+      //     assert.isNotNull(jpeg.barcode);
+      //   });
+      // });
+      // it('pdf barcodes should checksum sha16 successfully against base36 barcode component', () => {
+      //   decodedJpegs.forEach((jpeg) => {
+      //     const { base36, sha16 } = getBarcodeSegments(jpeg.barcode);
+      //     const recomputedSha16 = jsCrc.crc16(base36);
+      //     assert.equal(recomputedSha16, sha16);
+      //     validBarcodes.push(jpeg.barcode);
+      //   });
+      // });
+    });
+  });
+
+  describe('Barcode submission on `pl-barcode-scan` element (before pdf scan upload)', () => {
+    const studentCourseInstanceUrl = baseUrl + '/course_instance/1';
+
+    before('create students', async () => {
+      defaultUser = { authUid: config.authUid, authName: config.authName, authUin: config.authUin }; // test suite default
+      for (const student of mockStudents) {
+        setUser(student);
+        await fetch(baseUrl);
+      }
+    });
+    before('get homework 1 url', async () => {
+      setUser(mockStudents[0]);
+      const courseInstanceBody = await (await fetch(studentCourseInstanceUrl)).text();
+      const $courseInstancePage = cheerio.load(courseInstanceBody);
+      hm1AutomaticTestSuiteUrl =
+        baseUrl.replace('/pl', '') +
+        $courseInstancePage('a:contains("Homework for automatic test suite")').attr('href');
+    });
+    after('restore default user', () => {
+      setUser(defaultUser);
+    });
+
+    it('students should NOT be able to submit invalid barcodes', async () => {
+      for (const student of mockStudents) {
+        setUser(student);
+        const hm1BarcodeSubmissionUrl = await getBarcodeSubmissionUrl(
+          baseUrl,
+          hm1AutomaticTestSuiteUrl
+        );
+        // front-end validation doesn't work here, but we try backend validation from pl-barcode-scan.py
+        const save = await saveOrGrade(
+          hm1BarcodeSubmissionUrl,
+          { _pdf_barcode_scan: 9999999 },
+          'save'
+        );
+        assert.include(
+          await save.text(),
+          'Submitted answer\n          \n        </span>\n        <span>\n    \n        \n            <span class="badge badge-danger">invalid, not gradable</span>'
+        );
```

Comment:
The inclusion of whitespace in the assertion looks really brittle. Can we instead just check for the string "invalid, not gradable"? Ditto for the following tests.

Reply:
Yeah, this is kind of annoying, but there is a slight difference between a first and second submission.

I avoid resetting the database or creating another user to test with this. I think this is the simplest approach that I found.

![image](https://user-images.githubusercontent.com/7872295/147262949-8f0748eb-fb54-4d8a-bee5-7108294a9d6c.png)

Reply:
It does look a bit brittle, but it would take the template changing or the underlying logic somewhere to break this.

Reply:
I also don't like this though. I wish there was a better way to test this and make it more readable without introducing more complexity in the tests.

Reply:
You should be able to use [Cheerio](https://www.npmjs.com/package/cheerio) to more precisely query the DOM, we do that in a lot of other tests!